### DISCUSSION TITLE
Discuss about Valibot 🤖

### DISCUSSION THREAD

**Discussion by fabian-hiller** [open-circle/valibot#52] [Announcements]
Everything that does not fit to an issue or PR will probably find the right place here. We are happy about every new idea! Please be respectful to each other when discussing.

**Comment by jamesarosen** [comment]
It's not completely clear to me when [transformations](https://valibot.dev/guides/pipelines/#transformations) apply. I commonly need to map from something that has `entries()` to an `object`, so I'd like to have something like this:

```ts
type HasEntries<T = string, U = unknown> = {
  entries: () => [T, U][];
}

const entriesToObject = (x: HasEntries) => Object.fromEntries(x.entries);
```

Would this be a transform? How would I chain it together with an object schema like the following?

```ts
const mySchema = object({
  foo: optional(string()),
  bar: string(),
})
```

**Reply by fabian-hiller** [reply]
Since you plan to change the data type (if I understood you correctly) the `transform` method is the right choice: https://valibot.dev/guides/methods/#transform

**Reply by jamesarosen** [reply]
The only example of Transform in the docs shows the transform happening after the parse:

```
const StringSchema = transform(string(), (input) => input.length);
const stringOutput = parse(StringSchema, 'hello'); // 5
```

I want to transform the _input_ before the parse. This doesn't work:

```
const Lastchema = transform((input) => input.at(-1), string())
const lastString = parse(Lastchema, [1, 2, 'three'])
Uncaught TypeError: schema._parse is not a function
    at Object._parse (node_modules/valibot/dist/index.cjs:4027:29)
    at parse (node_modules/valibot/dist/index.cjs:3732:25)
```

**Reply by fabian-hiller** [reply]
Then `coerce` is the right choice: https://valibot.dev/guides/methods/#coerce

Note that you also can pass a schema as the last argument to `transform`. This way you can validate the input before and after the transformation.

**Reply by jamesarosen** [reply]
This next question is a bit more philosophical. `coerce` only accepts an `(value: unknown) => T`. Would you expect folks who use `coerce` to narrow the type via a first round of validation? That is, `validate is S-> coerce S to T -> parse T`.

The alternative would be to change `coerce` to `coerce<S, T>(value: S) => T`, which would then allow only an `S` to be passed to the resulting schema.

**Reply by fabian-hiller** [reply]
`coerce` is for forcing unknown data into a specific data type. If you want to transform validated data, you should use `transform`. The last argument of `transform` allows you to validate the data after the transformations, if necessary. Does that answer your question?

**Reply by jamesarosen** [reply]
Maybe, but only I think of that validation-after-transform as also having coercion.

Here's what I started with, without valibot:

```ts
function parseOptionalDate(s: string | null) {
  if (s == null) return null
  d = new Date(s)
  return isNaN(d) ? null : d
}

function parseRangeParams(searchParams: URLSearchParams) {
  return {
    from: parseOptionalDate(searchParams.get('from')),
    to: parseOptionalDate(searchParams.get('to')),
  }
}
```

I can valibot-ize it a bit like so:

```ts
const MaybeDateSchema = coerce(optional(date()), (i) => { d = new Date(i); return isNaN(d) ? null : d })

const RangeSchema = object({
  from: optional(date()),
  to: optional(date()),
})

function parseRangeParams(searchParams: URLSearchParams) {
  return parse(RangeSchema, {
    from: searchParams.get('from'),
    to: searchParams.get('to'),
  })
}
```

But I'm getting stuck on fully valibot-izing it so that I can do this:

```
// from: Date | null, to: Date | null, searchParams: URLSearchParams
const { from , to } = parse(RangeSearchParamsSchema, searchParams)
```

I feel like I need to `coerce` the `URLSearchParams` to an `Object`, but I want to "pipe" that into another schema afterwards.

**Reply by fabian-hiller** [reply]
I think in your use case it makes more sense to split it into two steps, as in the second code example. Squeezing everything into one schema makes the code unintelligible.
